File: src/main/java/io/github/oleksivio/telegram/bot/core/controller/annotations/parser/unpacker/CallbackQueryUnpackerParser.java
package io.github.oleksivio.telegram.bot.core.controller.annotations.parser.unpacker;

import io.github.oleksivio.telegram.bot.core.controller.handler.unpack.UpdateUnpacker;
import org.springframework.stereotype.Component;
import io.github.oleksivio.telegram.bot.api.annotations.receiver.CallbackQueryReceiver;
import io.github.oleksivio.telegram.bot.api.model.objects.std.CallbackQuery;

import java.util.Optional;

@Component
public class CallbackQueryUnpackerParser extends UnpackerParser<CallbackQueryReceiver, CallbackQuery> {

    @Override
    public Class<CallbackQueryReceiver> getAnnotationClass() {
        return CallbackQueryReceiver.class;
    }

    @Override
    public UpdateUnpacker<CallbackQuery> getUpdateUnpacker() {
        return new UpdateUnpacker<>(in -> Optional.ofNullable(in.getCallbackQuery()), CallbackQuery.class);
    }

}

File: src/main/kotlin/io/github/oleksivio/telegram/bot/core/controller/annotations/parser/unpacker/CallbackQueryUnpackerParser.kt
package io.github.oleksivio.telegram.bot.core.controller.annotations.parser.unpacker

import io.github.oleksivio.telegram.bot.api.annotations.receiver.CallbackQueryReceiver
import io.github.oleksivio.tl.kbot.server.api.objects.std.CallbackQuery
import io.github.oleksivio.telegram.bot.core.controller.handler.unpack.UpdateUnpacker
import org.springframework.stereotype.Component

@Component
class CallbackQueryUnpackerParser : UnpackerParser<CallbackQueryReceiver, CallbackQuery>() {
    override val updateUnpacker: UpdateUnpacker<CallbackQuery>
        get() = UpdateUnpacker({ it.callbackQuery }, CallbackQuery::class)

    override val parserAnnotationClass = CallbackQueryReceiver::class

}
